File: WolfDen/Game/Behaviours/RepairBehaviour/RepairBehaviour.cs
using SQLitePCL;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using WolfDen.Game.EventArguments;
using WolfDen.Game.Interfaces;
using WolfDen.Game.Other;
using WolfDen.Game.Tiles;

namespace WolfDen.Game.Behaviours.RepairBehaviour
{
    public class RepairBehaviour : IBehaviour
    {
        public string Class { get { return GetType().Name; } }
        public string Id { get; set; }
        public IHealth UnitHealth { get; set; }
        public Unit Unit { get; set; }
        public int FullRepairCost { get; set; }
        public RepairCoordinator RepairCoordinator { get; set; }
        
        public Game Game;

        public RepairBehaviour(Game game, Unit unit)
        {
            Game = game;
            var awPlayer = unit.Owner;
            RepairCoordinator = awPlayer.RepairCoordinator;
            RepairCoordinator.CalcRepairCostsEvent += CalcRepairCosts;
            RepairCoordinator.ExecuteRepairsEvent += RepairUnit;            
        }

        public void CalcRepairCosts()
        {
            int repairCosts = 0;
            int percentageToRepair = 0;
            GetRepairCosts(ref repairCosts, ref percentageToRepair);
            RepairCoordinator.TurnRepairCosts += 0;
        }

        public void GetRepairCosts(ref int repairCosts, ref int percentageToRepair)
        {
            repairCosts = 0;
            percentageToRepair = 0;
            if (Unit.Tile is Building building && building.Owner == Unit.Owner)
            {
                // Check how much to repair.
                var missingHealth = UnitHealth.MaxHealth - UnitHealth.CurrentHealth;

                // Max repair is 20%.
                percentageToRepair = Math.Min(20, missingHealth * 100 / UnitHealth.MaxHealth);

                // Check if owner has enough gold.
                repairCosts = percentageToRepair * FullRepairCost / 100;
            }
        }

        public void RepairUnit()
        {
            int repairCosts = 0;
            int percentageToRepair = 0;
            GetRepairCosts(ref repairCosts, ref percentageToRepair);
            var awOwner = Unit.Owner;
            awOwner.Gold -= repairCosts;
            UnitHealth.CurrentHealth += UnitHealth.MaxHealth * percentageToRepair / 100;
        }

        // Not used for this behaviour. Bad design, sad times.
        public bool ExecuteMove(dynamic data)
        {
            throw new NotImplementedException();
        }

        public void Destroy()
        {
            RepairCoordinator.CalcRepairCostsEvent -= CalcRepairCosts;
            RepairCoordinator.ExecuteRepairsEvent -= RepairUnit;
        }
    }
}

File: WolfDen/wwwroot/ts/Behaviours/RepairBehaviour/RepairBehaviour.ts
namespace TurnBasedBoardGame {
    export class RepairBehaviour implements IDestroy {
        Unit: Unit;
        Id: string;
        FullRepairCost: number;
        UnitHealth: ICombatBehaviour;
        Game: Game;
        Animation: TextDisappearAnimation;
        RepairCoordinator: RepairCoordinator;

        Initialize() {
            let self = this;
            this.RepairCoordinator = this.Unit.Owner.RepairCoordinator;
            this.RepairCoordinator.CalcRepairCostsEvent[this.Unit.Id + " " + this.Id] = () => { self.CalcRepairCosts(); };
            this.RepairCoordinator.ExecuteRepairsEvent[this.Unit.Id + " " + this.Id] = () => { self.RepairUnit(); };
        }

        CalcRepairCosts() {
            let repairCosts = this.GetRepairCosts();
            this.RepairCoordinator.TurnRepairCosts += repairCosts.goldNeeded;
        }

        GetRepairCosts() {
            if (this.Unit.Node && IsBuilding(this.Unit.Node) && (this.Unit.Node as Building).Owner == this.Unit.Owner && this.UnitHealth.CurrentHealth < this.UnitHealth.MaxHealth) {

                // Check how much to repair.
                let missingHealth = this.UnitHealth.MaxHealth - this.UnitHealth.CurrentHealth;

                // Max repair is 20%.
                let percentageToRepair = Math.floor(Math.min(20, missingHealth * 100 / this.UnitHealth.MaxHealth));

                // Check if owner has enough gold.
                let goldNeeded = Math.floor(percentageToRepair * this.FullRepairCost / 100);
                return { goldNeeded: goldNeeded, percentageToRepair: percentageToRepair };
            }
            return { goldNeeded: 0, percentageToRepair: 0 };
        }

        RepairUnit() {
            let awOwner = this.Unit.Owner;
            let repairCosts = this.GetRepairCosts();

            if (repairCosts.goldNeeded == 0 && repairCosts.percentageToRepair == 0)
                return;

            // Substract gold and repair the unit.
            awOwner.Gold -= repairCosts.goldNeeded;
            this.UnitHealth.SetCurrentHealth(this.UnitHealth.CurrentHealth + Math.floor(this.UnitHealth.MaxHealth * repairCosts.percentageToRepair / 100));

            // Update HUD
            var tile = this.Unit.Node as Tile;
            var textStyle = PixiTextStyle1();
            textStyle.fill = "#46eb34";
            this.Animation = new TextDisappearAnimation("+" + repairCosts.percentageToRepair.toString(), textStyle, tile.X * terrainWidth + 20, tile.Y * terrainHeight - 5, mainContainer);
        }


        Destroy(): void {
            delete this.RepairCoordinator.CalcRepairCostsEvent[this.Unit.Id + " " + this.Id];
            delete this.RepairCoordinator.ExecuteRepairsEvent[this.Unit.Id + " " + this.Id];
        }
}
}